### PR TITLE
Create master pipeline based on Jenkinsfile

### LINKED ISSUE #4893: Define Master and Releases Jenkines Pipeline in one File
Subtask of https://github.com/mesosphere/marathon/issues/4867

We should ditch most pipelines and have a master and releases pipelines based on a Jenkinsfile so that we maintain them in code.

### LINKED ISSUE #4924: Timeout Stages in Pipelines
Subtask of epic https://github.com/mesosphere/marathon/issues/4867.

Some times builds get stuck such as [this one](https://jenkins.mesosphere.com/service/jenkins/view/Marathon/job/marathon-unit-integration-tests-pushes/9143/). Jenkins should kill these jobs after a timeout.

### PR DESCRIPTION
**Pull request by jeschkies**
This defines a pipeline for `master`, `releases/*` and `pipelines/*` branches.

The `marathon-unit-integration-tests-pushes` should be removed once this is merged.

Closes #4893.
Closes #4924.

### CODE REVIEW

File: src/test/scala/mesosphere/marathon/integration/setup/ZookeeperServer.scala [hunk 1]
```diff
@@ -1,19 +1,18 @@
 package mesosphere.marathon.integration.setup
 
 import com.typesafe.scalalogging.StrictLogging
-import mesosphere.marathon.core.storage.store.impl.zk.{ NoRetryPolicy, RichCuratorFramework }
+import mesosphere.marathon.core.storage.store.impl.zk.{NoRetryPolicy, RichCuratorFramework}
```
Comment (janisz) [inline src/test/scala/mesosphere/marathon/integration/setup/ZookeeperServer.scala]: Format code

Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/integration/setup/ZookeeperServer.scala]: Thanks for noting. The branch should have only changes to the Jenkinsfile. I'll clean that up.

**Comment by aquamatthias** [comment]
Can we enumerate the steps, so the result is shown in order?

**Comment by aquamatthias** [comment]
Can we add test coverage reporting?

**Comment by jeschkies** [comment]
Here is the new ordering once we remove the old pipelines:

screen shot 2017-01-17 at 17 50 22

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            checkout scm
+            gitCommit = sh(returnStdout: true, script: 'git rev-parse HEAD').trim()
+            shortCommit = gitCommit.take(8)
+            currentBuild.displayName = "#${env.BUILD_NUMBER}: ${shortCommit}"
+        }
+        stage("Install Mesos") {
+            sh "sudo apt-get -y clean"
+            sh "sudo apt-get -y update"
+            sh "sudo apt-get install -y --force-yes --no-install-recommends curl"
+            sh """if grep -q MesosDebian \$WORKSPACE/project/Dependencies.scala; then
```
Comment (janisz) [inline Jenkinsfile]: This looks weird. I think Mesos version should be a parameter so test could be easily run on different Mesos version not only  one that is deckared as a dependency.

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        MESOS_VERSION=\$(sed -n 's/^.*MesosDebian = "\\(.*\\)"/\\1/p' <\$WORKSPACE/project/Dependencies.scala)
+        sudo apt-get install -y --force-yes --no-install-recommends mesos=\$MESOS_VERSION
+      else
+        MESOS_VERSION=\$(sed -n 's/^.*mesos=\\(.*\\)&&.*/\\1/p' <\$WORKSPACE/Dockerfile)
+        sudo apt-get install -y --force-yes --no-install-recommends mesos=\$MESOS_VERSION
```
Comment (janisz) [inline Jenkinsfile]: This line could be moved after the condition.

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      fi"""
+        }
+        stageWithCommitStatus("1. Compile") {
+          try {
+            withEnv(['RUN_DOCKER_INTEGRATION_TESTS=true', 'RUN_MESOS_INTEGRATION_TESTS=true']) {
```
Comment (janisz) [inline Jenkinsfile]: Are this env vars necessary for this step?

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+              sh "sudo -E sbt -Dsbt.log.format=false clean compile scapegoat"
```
Comment (janisz) [inline Jenkinsfile]: IMO compile should be separated step from lint.

File: Jenkinsfile (hunk 2 above)
Comment (jeschkies) [inline Jenkinsfile]: Good point. We could define a matrix with different versions in the future. However, this resembles what we have right now.

File: Jenkinsfile (hunk 4 above)
Comment (jeschkies) [inline Jenkinsfile]: Yes. Otherwise we don't run certain tasks.

File: Jenkinsfile (hunk 5 above)
Comment (jeschkies) [inline Jenkinsfile]: As far as I can tell we would compile twice then and thus slow down the process.

File: Jenkinsfile (hunk 3 above)
Comment (jeschkies) [inline Jenkinsfile]: Done. Thanks.

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+      else
+        MESOS_VERSION=\$(sed -n 's/^.*mesos=\\(.*\\)&&.*/\\1/p' <\$WORKSPACE/Dockerfile)
+      fi
+      sudo apt-get install -y --force-yes --no-install-recommends mesos=\$MESOS_VERSION
+      """
+        }
+        stageWithCommitStatus("1. Compile") {
+          try {
+            withEnv(['RUN_DOCKER_INTEGRATION_TESTS=true', 'RUN_MESOS_INTEGRATION_TESTS=true']) {
+              sh "sudo -E sbt -Dsbt.log.format=false clean compile scapegoat"
```
Comment (jasongilanfarr) [inline Jenkinsfile]: I'd like to see docs earlier in the cycle.

File: Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        stageWithCommitStatus("3. Test Integration") {
+          try {
+              timeout(time: 20, unit: 'MINUTES') {
+                withEnv(['RUN_DOCKER_INTEGRATION_TESTS=true', 'RUN_MESOS_INTEGRATION_TESTS=true']) {
+                   sh "sudo -E sbt -Dsbt.log.format=false integration:test mesos-simulation/integration:test"
+                }
+            }
+          } finally {
+            junit allowEmptyResults: true, testResults: 'target/test-reports/integration/**/*.xml'
+          }
```
Comment (jasongilanfarr) [inline Jenkinsfile]: Why not get another set of coverage data?

File: Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        }
+        stage("4. Create docs") {
+            sh "sudo -E sbt -Dsbt.log.format=false doc"
+        }
```
Comment (jasongilanfarr) [inline Jenkinsfile]: Should we publish binaries at this step?

File: Jenkinsfile (hunk 6 above)
Comment (jeschkies) [inline Jenkinsfile]: Sure. Done.

File: Jenkinsfile (hunk 8 above)
Comment (jeschkies) [inline Jenkinsfile]: Eventually. For now I'll just archive them.

Comment (jeschkies) [inline Jenkinsfile]: We probably have to compile a release here since `sbt coverage test` is creating annotated bytecode.

**Pull request merged**